### PR TITLE
cli: add start command

### LINKED ISSUE #1014: `ramen start` that starts all services on a server with basic configuration.
And prints some helps regarding user and how to connect using the GUI.

### PR DESCRIPTION
**Pull request by darlentar**
closes #1014

### CODE REVIEW

**Comment by darlentar** [comment]
Take remarks into account.
I do not modify test tough. I think that it may be better to add a new one instead, for regressions problems. WDYT ?

File: src/RamenCliCmd.ml [hunk 1]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+  fork_cont ServiceNames.gc (fun () ->
+    let dry_run = false in
+    gc conf dry_run del_ratio compress_older gc_loop daemonize
+       to_stdout to_syslog prefix_log_with_name ()
+  ) ;
+  fork_cont ServiceNames.archivist (fun () ->
+    archivist conf archivist_loop daemonize false allocs reconf
+              to_stdout to_syslog prefix_log_with_name smt_solver ()
+  ) ;
+  let rec loop () =
```
Comment (rixed) [inline src/RamenCliCmd.ml:1401]: replay_service is still missing

File: src/RamenCliCmd.ml [hunk 2]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+    if not (List.is_empty !pids) then
+    (match restart_on_EINTR Unix.wait () with
+     | (0, _status) -> !logger.info "All process has been stopped"
```
Comment (rixed) [inline src/RamenCliCmd.ml:1404]: man 2 wait

File: src/RamenCliCmd.ml [hunk 3]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+     | (pid, _status) ->
+        let proc_name = List.assoc pid !pids in
```
Comment (rixed) [inline src/RamenCliCmd.ml:1406]: service_name

File: src/RamenCliCmd.ml [hunk 4]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+        !logger.info "%a process (%d) has been stopped" N.service_print proc_name pid ;
```
Comment (rixed) [inline src/RamenCliCmd.ml:1407]: Prefer perfect form here (has stopped) or it sounds like it hasn't completed normally but rather have been terminated.
Actually, you should properly distinguish those cases. The status of a process you execed is a super important piece of information that you must not ignore (a bit like, remember, the actual error string of an HTTP response ;))
There is a printer for that somewhere: RamenHelpers.string_of_process_status.

File: src/RamenCliCmd.ml [hunk 5]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+               no_source_examples archive_total_size
+               archive_recall_cost oldest_site
+               gc_loop archivist_loop allocs reconf
+               del_ratio compress_older = fun () ->
+  let open Unix in
+  let sync_url = List.hd ports in
+  if not (String.starts_with sync_url "127.0.0.1:") then
+    failwith ("confserver can only be binded to 127.0.0.1: " ^ sync_url) ;
+  let conf = {conf with C.sync_url = sync_url} in
+  let pids = ref [] in
```
Comment (rixed) [inline src/RamenCliCmd.ml:1351]: Too much lisp lately? You want a set here. :)

File: src/RamenHelpers.ml [hunk 6]
```diff
@@ -405,7 +405,9 @@ let rec restart_on_failure ?(while_=always) what f x =
   else
     try f x
     with e ->
-      print_exception e ;
+      match e with
```
Comment (rixed) [inline src/RamenHelpers.ml:408]: `with` can do pattern matching.

File: src/RamenHelpers.ml [hunk 7]
```diff
@@ -405,7 +405,9 @@ let rec restart_on_failure ?(while_=always) what f x =
...
+        | Stdlib.Exit -> ()
```
Comment (rixed) [inline src/RamenHelpers.ml:409]: Why is this Stdlib required, all of a sudden?

File: src/ramen.ml [hunk 8]
```diff
@@ -1267,6 +1267,49 @@ let archivist =
...
+
+let gc_loop =
+  let i = Arg.info ~doc:CliInfo.loop
+                   ["gc-loop"] in
+  Arg.(value (opt (some float) (Some Default.gc_loop) i))
+
+let archivist_loop =
+  let i = Arg.info ~doc:CliInfo.loop
+                   ["archivist-loop"] in
+  Arg.(value (opt (some float) (Some Default.archivist_loop) i))
```
Comment (rixed) [inline src/ramen.ml:1282]: Why this change? It's still nice to have the loop period optional.

File: src/RamenCliCmd.ml [hunk 9]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+     | 0 ->
+        child ()
+     | pid ->
+        pids := (pid, proc_name)::!pids;
+        !logger.info "Start %a process (%d)" N.service_print proc_name pid in
+  flush_all () ;
+  if daemonize then do_daemonize () ;
+  let daemonize = false in
+  let prefix_log_with_name = true in
+  (* TODO we check twice params here *)
```
[the remark this answers is not in the document]
Comment (darlentar) [reply src/RamenCliCmd.ml:1363]: No. It's just a code smell, so I wrote this sentence.

File: src/RamenCliCmd.ml [hunk 10]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+               archive_recall_cost oldest_site
+               gc_loop archivist_loop allocs reconf
+               del_ratio compress_older = fun () ->
+  let open Unix in
+  let sync_url = List.hd ports in
+  if not (String.starts_with sync_url "127.0.0.1:") then
+    failwith ("confserver can only be binded to 127.0.0.1: " ^ sync_url) ;
+  let conf = {conf with C.sync_url = sync_url} in
+  let pids = ref [] in
+  let fork_cont proc_name child =
```
[the remark this answers is not in the document]
Comment (darlentar) [reply src/RamenCliCmd.ml:1352]: Right

File: src/RamenCliCmd.ml [hunk 11]
```diff
@@ -1333,19 +1323,95 @@ let graphite_expand conf for_render since until query () =
...
+  fork_cont ServiceNames.confserver (fun () ->
+    confserver conf daemonize to_stdout to_syslog prefix_log_with_name ports
+               ports_sec srv_pub_key_file srv_priv_key_file no_source_examples
+               archive_total_size archive_recall_cost oldest_site ()
+  ) ;
+  fork_cont ServiceNames.choreographer (fun () ->
+    choreographer conf daemonize to_stdout to_syslog prefix_log_with_name ()
+  ) ;
+  fork_cont ServiceNames.execompserver (fun () ->
+    let max_simult_compilations = 10 in
```
[the remark this answers is not in the document]
Comment (darlentar) [reply src/RamenCliCmd.ml:1379]: It's already the case. This shouldn't be there though.

File: src/ramen.ml (hunk 8 above)
Comment (darlentar) [reply src/ramen.ml:1282]: IMHO, for the start command loop should be mandatory.

File: src/RamenCliCmd.ml (hunk 5 above)
Comment (darlentar) [reply src/RamenCliCmd.ml:1351]: Why ? This is a small list. And I need the association part that Set doesn't provide (I may use a Map though).

Comment (rixed) [reply src/RamenCliCmd.ml:1350]: You do not really need the association and could use a set of pid*service_name but of course a map would be OK too. Probably preferable indeed.

File: src/ramen.ml (hunk 8 above)
Comment (rixed) [reply src/ramen.ml:1282]: what's the rationale?

Comment (darlentar) [reply src/ramen.ml:1282]: That `ramen start` starts stuffs, and thus a run once program is a non-sense.

Comment (rixed) [reply src/ramen.ml:1282]: Yes you are right of course. In that case it makes no sense to keep the value optional.

File: src/RamenCliCmd.ml [hunk 12]
```diff
@@ -1333,19 +1322,102 @@ let graphite_expand conf for_render since until query () =
...
+  ) ;
+  let rec loop () =
+    if not (Map.Int.is_empty !pids) then
+    (match restart_on_EINTR Unix.wait () with
+     | (-1, status) -> !logger.info "All processes has stopped: %s"
+        (RamenHelpers.string_of_process_status status);
+        Processes.quit := Some 0
+     | (pid, status) ->
+        let service_name, _pids = Map.Int.extract pid !pids in
+        pids := _pids ;
```
Comment (rixed) [inline src/RamenCliCmd.ml:1411]: Underscore prefixess have a special meaning in OCaml: they mean the variable is not useful and therefore the compiler should not warn when they are unused. Unfortunately, it does not warn when a variable prefixed with undescrire is in fact used.
Better call this `_pids` differently. When I want such a short lived alias without shadowing, I personnaly use an underscore _suffix_, so here `pids_`.

File: src/RamenCliCmd.ml [hunk 13]
```diff
@@ -1333,19 +1322,102 @@ let graphite_expand conf for_render since until query () =
...
+  ) ;
+  fork_cont ServiceNames.replayer (fun () ->
+    replay_service conf daemonize to_stdout to_syslog prefix_log_with_name ()
+  ) ;
+  let rec loop () =
+    if not (Map.Int.is_empty !pids) then
+    (match restart_on_EINTR Unix.wait () with
+     | (-1, status) -> !logger.info "All processes has stopped: %s"
+        (RamenHelpers.string_of_process_status status);
+        Processes.quit := Some 0
```
Comment (rixed) [inline src/RamenCliCmd.ml:1408]: man 2 wait _and_ have a look at the OCaml doc for Unix wait. That's not how errors are reported.

**Comment by rixed** [comment]
Still the wait has to be fixed (errors are returned with exceptions, and there is no magical return value meaning "all the pids are terminated", and there cannot be if you think about it), and signals cascaded, and that new cli command used in the cli tests.

**Comment by darlentar** [comment]
> Still the wait has to be fixed (errors are returned with exceptions, and there is no magical return value meaning "all the pids are terminated", and there cannot be if you think about it), and signals cascaded, and that new cli command used in the cli tests.

You right of course. This code is a non sense.

**Comment by rixed** [comment]
Merged.
